src/aggregators/rdirect.jl

```diff
@@ -0,0 +1,136 @@
+  p.next_jump_time == t
+end
+
+# executing jump at the next jump time
+function (p::RDirectJumpAggregation)(integrator)
+  execute_jumps!(p, integrator, integrator.u, integrator.p, integrator.t)
+  generate_jumps!(p, integrator, integrator.u, integrator.p, integrator.t)
+  register_next_jump_time!(integrator, p, integrator.t)
+  nothing
+end
+
+# setting up a new simulation
+function (p::RDirectJumpAggregation)(dj, u, t, integrator) # initialize
+  initialize!(p, integrator, u, integrator.p, t)
+  register_next_jump_time!(integrator, p, t)
+  nothing
+end
+
+############################# Required Functions #############################
+
+# creating the JumpAggregation structure (tuple-based constant jumps)
+function aggregate(aggregator::RDirect, u, p, t, end_time, constant_jumps,
+                    ma_jumps, save_positions, rng; kwargs...)
+
+    # handle constant jumps using function wrappers
+    rates, affects! = get_jump_info_fwrappers(u, p, t, constant_jumps)
+
+    build_jump_aggregation(RDirectJumpAggregation, u, p, t, end_time, ma_jumps,
+                          rates, affects!, save_positions, rng; num_specs = length(u), kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::RDirectJumpAggregation, integrator, u, params, t)
+    fill_rates_and_sum!(p, u, params, t)
+    p.max_rate = maximum(p.cur_rates)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+"execute one jump, changing the system state and updating rates"
+function execute_jumps!(p::RDirectJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u)
+
+    # update rates
+    update_dependent_rates!(p, u, params, t)
+    nothing
+end
+
+"calculate the next jump / jump time"
+function generate_jumps!(p::RDirectJumpAggregation, integrator, u, params, t)
+    # if no more events possible there is nothing to do
+    sum_rate = p.sum_rate
+    if nomorejumps!(p, sum_rate)
+        return nothing
+    end
+    @unpack rng, cur_rates, max_rate = p
+
+    num_rxs = length(cur_rates)
+    rx = trunc(Integer, rand(rng) * num_rxs)+1
+    while cur_rates[rx] < rand(rng) * max_rate
+        rx = trunc(Integer, rand(rng) * num_rxs)+1
+    end
+
+    p.next_jump = rx
+
+    p.next_jump_time = t + randexp(p.rng) / sum_rate
+    nothing
+end
+
+
+######################## SSA specific helper routines #########################
+function update_dependent_rates!(p::RDirectJumpAggregation, u, params, t)
+    @inbounds dep_rxs = p.dep_gr[p.next_jump]
+    @unpack ma_jumps, rates, cur_rates, sum_rate = p
+    need_to_reculculate_max_rate = false
+    @inbounds for rx in dep_rxs
+        sum_rate -= cur_rates[rx]
+        @inbounds new_rate = calculate_jump_rate(ma_jumps, rates, u,params,t,rx)
+        sum_rate += new_rate
+        if new_rate > p.max_rate
+            p.max_rate = new_rate
+        elseif abs(cur_rates[rx]-p.max_rate) < eps(typeof(p.max_rate))
```

Comment:
What is the point of this check? Isn't `need_to_reculculate_max_rate` already `false`? In fact, do you ever set it to `true` in this function?

Reply:
It was supposed to be `need_to_recalculate_max_rate = true`. The idea is that if the rate that is equal to the current max_rate decreased, then we need to recalculate max_rate. This keeps max_rate always equal to the current maximum rate.

Reply:
Can't you just do `==` here, since they really should be equal?